File: backend/rpc/handler/system/backup.py
from flask import request

from rpc.gen.system.backup.services import TSystemBackupService
from rpc.gen.user.user.types.ttypes import TUserRole
from rpc.gen.system.backup.structs.ttypes import TBackupFile

import db
from models import models

class TSystemBackupServiceHandler(TSystemBackupService.Iface):
    def __init__(self):
        self.auth_model = models['auth']
        self.backup_model = models['backup']

    def create_backup(self, auth_token, name):
        auth_payload = self.auth_model.require_role(auth_token, TUserRole.ADMIN_UTAMA)
        file_name, last_modified = self.backup_model.create_backup(name)
        return TBackupFile(file_name, last_modified)

    def fetch_backups(self, auth_token):
        auth_payload = self.auth_model.require_role(auth_token, TUserRole.ADMIN_UTAMA)
        files = self.backup_model.fetch_backups()
        return [TBackupFile(file_name, last_modified) for file_name, last_modified in files]

    def get_download_token(self, auth_token, file_name):
        ip = request.remote_addr
        return self.backup_model.create_download_token(ip, file_name)

    def get_upload_token(self, auth_token, file_name):
        ip = request.remote_addr
        return self.backup_model.create_upload_token(ip, file_name)

    def delete_backup(self, auth_token, file_name):
        auth_payload = self.auth_model.require_role(auth_token, TUserRole.ADMIN_UTAMA)
        return self.backup_model.delete_backup(file_name)

    def restore_backup(self, auth_token, file_name):
        auth_payload = self.auth_model.require_role(auth_token, TUserRole.SUPER_ADMIN)
        self.backup_model.restore_backup(file_name)
        db.commit()

File: frontend/src/rpc/client/system/backup.js
import TSystemBackupService from '@/rpc/gen/TSystemBackupService';
import { TBaseClient } from '@/rpc/client/base';
import { TUserRole } from '@/rpc/gen/user.user.types_types';
import { TFileError, TFileErrorCode } from "@/rpc/gen/file.file.errors_types";
import { TUploadError, TUploadErrorCode } from "@/rpc/gen/file.upload.errors_types";
import axios from 'axios';


class TSystemBackupClient extends TBaseClient{
	constructor(stores=null){
		super(stores, TSystemBackupService, '/api/system/backup');
		return this;
	}

	async fetch_backups(){
		this.stores.helper.auth.requireRole(TUserRole.ADMIN_UTAMA);
		return await this.client.fetch_backups(this.stores.auth.authToken);
	}
	async create_backup(name){
		this.stores.helper.auth.requireRole(TUserRole.ADMIN_UTAMA);
		return await this.client.create_backup(this.stores.auth.authToken, name);
	}
	async delete_backup(file_name){
		this.stores.helper.auth.requireRole(TUserRole.ADMIN_UTAMA);
		return await this.client.delete_backup(this.stores.auth.authToken, file_name);
	}
	async get_download_token(file_name){
		this.stores.helper.auth.requireRole(TUserRole.ADMIN_UTAMA);
		return await this.client.get_download_token(this.stores.auth.authToken, file_name);
	}
	async download_backup(file_name){
		if (!file_name) throw new TFileError({ code: TFileErrorCode.FILE_NAME_EMPTY});
		var token = await this.get_download_token(file_name);
		var response = await axios({
			method: 'post',
			url: defaultBackendUrl + "/backup/download/",
			data: { download_token: token },
			responseType: 'arraybuffer'
		});
		let blob = new Blob(
			[response.data], 
			{ type: 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet' }
		);
		return blob;
	}
	async get_upload_token(file_name){
		this.stores.helper.auth.requireRole(TUserRole.ADMIN_UTAMA);
		return await this.client.get_upload_token(this.stores.auth.authToken, file_name);
	}
	async upload_backup(file, file_name=null){
		if (!file) throw new TUploadError({ code: TUploadErrorCode.FILE_NOT_PROVIDED});
		file_name = file_name || file.name;
		if (!file_name) throw new TUploadError({ code: TUploadErrorCode.FILE_NOT_PROVIDED});
		//if (!file.accepted) throw new TUploadError({ code: TUploadErrorCode.FILE_INVALID});
		var token = await this.get_upload_token(file_name);

		const form = new FormData();
		form.append("file", file);
		form.append("upload_token", token);

		var response = await axios({
			method: 'post',
			url: defaultBackendUrl + "/backup/upload/",
			data: form,
			headers: {'Content-Type': 'multipart/form-data' },
			responseType: 'json'
		});
		return response.data;
	}
	async restore_backup(file_name){
		console.log("File name: " + file_name);
		await this.client.restore_backup(this.stores.auth.authToken, file_name);
		await this.stores.client.auth.logout();
		await this.stores.app.setGlobalRefresh(true);
	}
}

const systemBackupClient = new TSystemBackupClient();

export { TSystemBackupClient, systemBackupClient }
export default systemBackupClient